rust/lance/src/io/exec/scalar_index.rs

```diff
@@ -0,0 +1,325 @@
+    pub fn new(dataset: Arc<Dataset>, expr: ScalarIndexExpr) -> Self {
+        Self { dataset, expr }
+    }
+
+    async fn do_execute(expr: ScalarIndexExpr, dataset: Arc<Dataset>) -> Result<RecordBatch> {
+        let query_result = expr.evaluate(dataset.as_ref()).await?;
+        let query_result_arr = query_result.into_arrow()?;
+        Ok(RecordBatch::try_new(
+            SCALAR_INDEX_SCHEMA.clone(),
+            vec![query_result_arr],
+        )?)
+    }
+}
+
+#[pin_project]
+struct StreamWithSchema {
+    #[pin]
+    stream: BoxStream<'static, datafusion::common::Result<RecordBatch>>,
+    schema: Arc<Schema>,
+}
+
+impl Stream for StreamWithSchema {
+    type Item = datafusion::common::Result<RecordBatch>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let this = self.project();
+        this.stream.poll_next(cx)
+    }
+}
+
+impl RecordBatchStream for StreamWithSchema {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+impl ExecutionPlan for ScalarIndexExec {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        SCALAR_INDEX_SCHEMA.clone()
+    }
+
+    fn output_partitioning(&self) -> datafusion::physical_plan::Partitioning {
+        Partitioning::RoundRobinBatch(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[datafusion::physical_expr::PhysicalSortExpr]> {
+        // No guarantee a scalar index scan will return row ids in any meaningful order
+        None
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        todo!()
+    }
+
+    fn execute(
+        &self,
+        _partition: usize,
+        _context: Arc<datafusion::execution::context::TaskContext>,
+    ) -> datafusion::error::Result<datafusion::physical_plan::SendableRecordBatchStream> {
+        let batch_fut = Self::do_execute(self.expr.clone(), self.dataset.clone());
+        let stream = futures::stream::iter(vec![batch_fut])
+            .then(|batch_fut| batch_fut.map_err(|err| err.into()))
+            .boxed()
+            as BoxStream<'static, datafusion::common::Result<RecordBatch>>;
+        Ok(Box::pin(StreamWithSchema {
+            schema: SCALAR_INDEX_SCHEMA.clone(),
+            stream,
+        }))
+    }
+
+    fn statistics(&self) -> datafusion::physical_plan::Statistics {
+        todo!()
+    }
+}
+
+lazy_static::lazy_static! {
+    pub static ref MATERIALIZE_INDEX_SCHEMA: SchemaRef = Arc::new(Schema::new(vec![ROW_ID_FIELD.clone()]));
+}
+
+/// An execution node that performs a scalar index search and materializes the mask into row ids
+///
+/// First, the index is searched to determine the mask that should be applied.  Then, we take the
```

Comment:
Materialize as we materialize / split it to disk? or just sth in memory as list of row ids?

Reply:
In memory as a list of row ids.